sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/QueueAsyncClient.java

```diff
@@ -704,16 +811,71 @@ public HttpPipeline getHttpPipeline() {
             marker -> StorageImplUtils.applyOptionalTimeout(this.client.getMessages()
                 .dequeueWithResponseAsync(queueName, maxMessages, visibilityTimeoutInSeconds,
                     null, null, context), timeout)
-                .map(response -> new PagedResponseBase<>(response.getRequest(),
-                    response.getStatusCode(),
-                    response.getHeaders(),
-                    response.getValue(),
-                    null,
-                    response.getDeserializedHeaders()));
+                .flatMap(this::transformMessagesDequeueResponse);
 
         return new PagedFlux<>(() -> retriever.apply(null), retriever);
     }
 
+    private Mono<PagedResponseBase<MessagesDequeueHeaders, QueueMessageItem>> transformMessagesDequeueResponse(
+        MessagesDequeueResponse response) {
+        List<QueueMessageItemInternal> queueMessageInternalItems = response.getValue();
+        List<QueueMessageItem> queueMessageItems = Collections.emptyList();
+        Mono<Void> mono = Mono.empty();
+        if (queueMessageInternalItems != null) {
+            queueMessageItems = new ArrayList<>(queueMessageInternalItems.size());
+            for (QueueMessageItemInternal queueMessageItemInternal : queueMessageInternalItems) {
+                try {
+                    queueMessageItems.add(transformQueueMessageItemInternal(queueMessageItemInternal, messageEncoding));
+                } catch (IllegalArgumentException e) {
+                    if (messageDecodingFailedHandler != null) {
+                        mono = mono.then(messageDecodingFailedHandler.apply(
```

Comment:
Does this failed handler return an error or just log a message?

Reply:
it's customer provided so we expect that error handling is implemented inside that handler. However if they get it wrong we want to surface errors in receive/peek API call result (like we do in .NET).
See `Peek with handler exception` test.